### PR TITLE
feat: support parsing for parquet writer option

### PR DESCRIPTION
**Pull request by fansehep**
# Which issue does this PR close?

part https://github.com/apache/arrow-rs/issues/4693

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by tustvold — COMMENTED** [review]
Could we get some tests for this please :pray:

**Comment by fansehep** [comment]
> Could we get some tests for this please 🙏

🫣i will add some tests later.

File: parquet/src/basic.rs [hunk 1]
```diff
@@ -2130,4 +2235,93 @@ mod tests {
...
+
+        // test unknown string
+        match "plain_xxx".parse::<Encoding>() {
+            Ok(e) => {
+                panic!("Should not be able to parse {:?}", e);
+            }
+            Err(e) => {
+                assert_eq!(e.to_string(), "Parquet error: unknown encoding: plain_xxx");
+            }
+        }
```
Comment (tustvold) [inline parquet/src/basic.rs:2272]:
```suggestion
        let err = "plain_xxx".parse::<Encoding>().unwrap_err();
        assert_eq!(e.to_string(), "Parquet error: unknown encoding: plain_xxx");
```

File: parquet/src/basic.rs [hunk 2]
```diff
@@ -2130,4 +2235,93 @@ mod tests {
...
+        compress = "brotli(3)".parse().unwrap();
+        assert_eq!(
+            compress,
+            Compression::BROTLI(BrotliLevel::try_new(3).unwrap())
+        );
+        compress = "lz4".parse().unwrap();
+        assert_eq!(compress, Compression::LZ4);
+
+        // test unknown compression
+        match "unknown".parse::<Compression>() {
```
Comment (tustvold) [inline parquet/src/basic.rs:2302]: These could also make use of unwrap_err()

File: parquet/src/basic.rs [hunk 3]
```diff
@@ -278,6 +279,25 @@ pub enum Encoding {
     BYTE_STREAM_SPLIT,
 }
 
+impl FromStr for Encoding {
+    type Err = ParquetError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_owned().to_uppercase().as_str() {
```
Comment (tustvold) [inline parquet/src/basic.rs:286]: I wonder if we should be case sensitive and leave it to the caller to force to upper case if that is what they want?

It seems strange that we would support parsing things like `PlAiN`

**Review by tustvold — APPROVED** [review]
Mostly just some minor test nits, but I also wonder if we really want to be case insensitive?

I think it probably also makes sense to derive Display if we are going to define FromStr, but happy for this to be a separate PR.

File: parquet/src/basic.rs (hunk 3 above)
Comment (fansehep) [reply parquet/src/basic.rs:286]: > It seems strange that we would support parsing things like `PlAiN`

For users, whose usage is generally unobservable, I also find it odd to support this kind of parsing like `PlAiN`, but it seems that both cases should be supported?

Comment (tustvold) [reply parquet/src/basic.rs:286]: I think we should just support case sensitive parsing and the user can opt in to case insensitive parsing by converting the input to uppercase should they want this behaviour

Comment (fansehep) [reply parquet/src/basic.rs]: > I think we should just support case sensitive parsing and the user can opt in to case insensitive parsing by converting the input to uppercase should they want this behaviour

done. i think it's a good suggestion.

**Comment by tustvold** [comment]
Thank you

**Pull request merged**